### DISCUSSION TITLE
Reworking how time works in BioCro

### DISCUSSION THREAD

**Discussion by leavesandlemmas** [biocro/biocro#136] [Ideas]
Cross post of the discussion on pull request #135 and issue #129.

There are also some other comments on another discussion: https://github.com/orgs/biocro/discussions/24

**Comment by eloch216** [comment]
I'll try to summarize the conclusions from today's meeting about reowkring BioCro's time system:

1. In the short term, the priority is to check to make sure the `time` column of the `drivers` input to `run_biocro` is ordered and evenly spaced. Scott O. has already worked on this; see PR https://github.com/biocro/biocro/pull/137.

2. In the long term, there are several planned changes. I think they can be accomplished in this order:

2a. Move the `timestep` input to the `ode_solver` input rather than having it as a parameter, since it is not a property of the model itself.

2b. Allow the user to (optionally) specify a `start_time` and `end_time` as part of the `ode_solver` input so a model can be run without drivers.

2c. Switch the time basis to values of `time` rather than indices of the `drivers` table.

In the even longer term, 2c would enable us to deal with non-equally-spaced drivers.

We may also consider providing other `boost` solvers based on the `integrate_adaptive` function, which would automatically determine the time points included in the solution.

Another change that is somewhat independent of the others is to remove any code from `run_biocro` that is specific to `doy` and `hour`.

Any changes to the `run_biocro` inputs should be accompanied by sensible defaults to minimize disruption to scripts written for older versions of BioCro.

(Edit 1: Changed `elapsed_time` to `time`.)

**Reply by eloch216** [reply]
I've also attached the slides I shared during the meeting

[2025-01-06 Time in BioCro.pptx](https://github.com/user-attachments/files/18322412/2025-01-06.Time.in.BioCro.pptx)

**Comment by justinmcgrath** [comment]
As a small point, `elapsed_time` as used here is what the time is in the dynamical system. Calling it simply `time` would match that more closely.

I was wondering about Scott R's comment about names for how we're handling time as discrete vs continuous. From what I gather, a "dynamical system" in general only requires that time be a monoid, which covers discrete times. If you want continuous times, I think the requirement is that time be a field -- roughly, a set with multiplication, addition, and inversibility for both operations. In that case they call them continuous dynamical systems or flows. That is how I've been thinking of this, rather than only as discrete systems.

**Reply by eloch216** [reply]
I changed `elapsed_time` to `time` in the comment above.

**Reply by leavesandlemmas** [reply]
>  From what I gather, a "dynamical system" in general only requires that time be a monoid, which covers discrete times. If you want continuous times, I think the requirement is that time be a field -- roughly, a set with multiplication, addition, and inversibility for both operations.

A field isn't necessary, although time is modeled using real numbers (which form a field). It's not really the fact that time is a monoid or group that matters, dynamical systems have a monoid/group action (which doesn't form a field). The difference between the group and monoid action is whether or not the dynamics are time-reversible (group and monoid if yes, monoid only if no). Say there's an initial state `x`, and a function `f(t, x) = y` which maps `x` to the state  `y` after `t` units of time have passed; `f` is basically the solution of the ODE system. The group/monoid action is the fact that the state `x` goes to `z` after `s + t` time  (`z = f(s + t, x)` ) then that's where `y` goes after `s` time has passed, `z=f(s, y)` where `y = f(t,x)` : i.e.,  `f(s + t, x) = f(s, f(t, x))` .  So the addition of times is the same as the composition of `f`. 

It's the action of a monoid because there might be multiple initial states that map to the same final state; the function `f(t, x)` is not invertible in `x` (AKA time isn't reversible). Continuous time dynamical systems are usually the [flow](https://en.wikipedia.org/wiki/Flow_(mathematics)) of a vector field, and they usually form a group action, where the composition of flows is the same as adding times, but they are invertible (at least locally). Basically if `f(t,x) = y` then `f(-t, y) = x`. Discrete time dynamical systems can be non-invertible (like the logistic map, tent map, etc.).  

I think the key difference between `elapsed_time` and the drivers is that there are sort of two kinds of time because most of the crop models are non-autonomous differential equations so the starting calendar time matters because some of the drivers depend on the calendar time. Because changing the starting date and time of a simulation would change the environmental variables, the calendar time is effectively like another state variable. However, calendar time and elapsed time are related because if `t` is the calendar time and `s` is the `elapsed_time` then `t = t_0 + s`.  Some of my modules (and some like the `harmonic_oscillator` ones) don't use any sort of calendar time since they don't have drivers.  Maybe we could use this difference to inform the naming? 

Ed's right that calendar times (or datetimes) don't form a group or monoid; they can't really be added per se; and there's no natural zero. But there is a natural group action acting on calendar times /datetimes and it's basically the map `t = t_0 + s`  where `s` is a duration of time (which does form a group). For instance, if `t_0 = 2025-01-01 14:00  CST`  and `s = 314 h` then the group action sends `t_0` to `t = 2025-01-14 16:00 CST`.

**Comment by leavesandlemmas** [comment]
Regarding @gsrohde 's question about biocro simulations for the southern hemisphere (see biocro/discussion rather than biocro/biocro/discussion), there doesn't seem to be any issue with running biocro  for multiple years if the dataframe spans multiple years. The `time` variable behaves strangely because it doesn't reflect the year, just the day of year plus a fraction of each day.  It also jumps 24 hours at each new year. The biological behavior isn't sensible but then again the model for miscanthus wasn't built to run several years I think. 

If we have a `time` variable, it shouldn't be defined as `time = doy  + hour / 24` because that would cause problems for any simulation spanning a new year. Maybe it should depend on the year.  Or maybe include an explict modulo. 

 ```r 
time = 365.2425  *year +  doy + hour/24 
# or 
time = (doy + hour/24) %% (365 + is_leap_year(year) 

is_leap_year <- function(year){
  divisible_by_4 = (year %% 4 == 0) 
  divisible_by_100 = (year %% 100 == 0)
  divisible_by_400 = (year %% 400 == 0) 
  return (divisible_by_4 & (!divisible_by_100 | divisible_by_400))   
}
```


Here's the code I used to run BioCro for three years. 
 ```
#  BioCro version 3.1.3 develop
library(BioCro)

three_year_weather <-  weather[c( '2005', '2006', '2007' )]
three_year_weather <- dplyr::bind_rows(three_year_weather, .id = 'year') |> as.data.frame()
three_year_weather$year <- three_year_weather$year |> as.numeric()

result <- run_biocro(
  miscanthus_x_giganteus$initial_values,
  within(miscanthus_x_giganteus$parameters, sowing_time <- 150),
  three_year_weather,
  miscanthus_x_giganteus$direct_modules,
  miscanthus_x_giganteus$differential_modules,
  miscanthus_x_giganteus$ode_solver
)

starting_date = paste0(result$year[1],"-01-01 00:00:00") |> as.POSIXct()
dt = seq(0, length=nrow(result), by=1) |> as.difftime(units='hours')

result$datetime <- starting_date + dt


library(ggplot2)
ggplot(result, aes(datetime, canopy_assimilation_rate)) + theme_bw() + geom_line()
```

**Comment by leavesandlemmas** [comment]
Regarding 2c, here's a prototype function to identify the proper row index from a list of times. I put a python implementation first to act like pseudo code.

```python
def find_interval_index(time : float, times : list) -> int:
    times.sort()
    first, last = times[0], times[-1]
    outside = (time < first) or (time > last)
    if outside:
        raise ValueError('`time` outside domain of `times`.')
    max_index = len(times) - 1
    for i in range(max_index):
        if times[i] <= time < times[i+1]:
            return i
    return max_index - 1

```

C ++ version. 
```cpp
size_t find_interval_index(double time, std::vector<double> times){
    std::sort(times.begin(), times.end());
    
    size_t const max_index {times.size() - 1};
    double const first = times[0];
    double const last = times[max_index];

    bool outside  = (time < first) || (time > last);
    if (outside) {
        throw std::invalid_argument("`time` outside domain of `times`.");
        return 0; // so that function return type is pure
    }  

    for (size_t i = 0; i < max_index; ++i){
        bool in_interval = (times[i] <= time) && (time < times[i+1]);
        if (in_interval) {
            return i;
        }
    }
    
    return max_index - 1; // use last interval if time equals last entry of times.
}

```

**Reply by leavesandlemmas** [reply]
I don't really understand the structure of the `dynamical_system` class in the biocro framework code, but it would be possible to have uneven drivers using a function-object class in c++, which wrap a list of driver values and times into a function which returns the driver at any time.

I didn't write a constructor but this gives the idea. The constructor should sort the abscissa, and then the `find_interval_index` function wouldn't need to do that at every time point. 


```cpp
class driver_function {
    using driver = std::unordered_map<std::string, double>;
    std::vector<double> abscissa;
    std::vector<driver> drivers_at_abscissa;

    // alternative
    // using timeseries = std::vector<double>;
    // timeseries abscissa;
    // std::unordered_map<std::string, timeseries> drivers_at_abscissa;

    public:

    driver operator()(double time){
        size_t idx = find_interval_index(time, abscissa);
        double interval_length = abscissa[idx+1] - abscissa[idx];
        double u = (time - abscissa[idx])/interval_length;
        driver out;

        std::transform( 
            drivers_at_abscissa[idx].cbegin(),
            drivers_at_abscissa[idx].cend(),
            drivers_at_abscissa[idx+1].cbegin(),
            out.begin(),            
            double [u](double x, double y){(1 -u) * x  +  u  *  y}
            );

        return out;
    }
    
};
```

**Reply by justinmcgrath** [reply]
Find a location in a sorted list is the the sort of problem that I am sure computer scientists have studied exhaustively. I would check what's out there and see if there's something that implements it already in the STL, something simple to copy as a utility function of our own, or something in Boost, with preference in that order. It's probably available for the ordered container types.

**Reply by leavesandlemmas** [reply]
A binary search would be faster. Here's an implemention using more of the STL. 

```cpp
double find_interval_index(double time, std::vector<double> times){
  auto lower =  std::lower_bound(times.cbegin(), times.cend(), time);
  return std::distance(times.cbegin(), lower);
 }
```
Also the `std::lower_bound` returns an iterator, which can be used to access the element. There's also a `std::upper_bound` function.

I tested the code using this browser [c++ shell](https://cpp.sh/)
```cpp
#include <stdexcept>
#include <iostream>
#include <vector>
#include <algorithm>

double find_interval_index(double time, std::vector<double> times){
  auto lower =  std::lower_bound(times.cbegin(), times.cend(), time);
  return std::distance(times.cbegin(), lower);
 }


int main(){
    std::vector<double> times{0.0, 1.0, 0.5, 2.0, 3.0};
    std::sort(times.begin(), times.end()); 
    
    std::cout << find_interval_index(0.0, times) << '\n';
    std::cout << find_interval_index(0.1, times) << '\n';
    
    std::cout << find_interval_index(0.55, times) << '\n';
    std::cout << find_interval_index(1.5, times) << '\n';
    std::cout << find_interval_index(2.9, times) << '\n';
    std::cout << find_interval_index(3.0, times) << '\n';
    

}
```
